[OSS101] Task 2: List all adjacent nodes of the hovered node in collaboration networks for quick reference

wouldmm opened the pull request:
新建了一个文件夹用于新特性，鼠标放置能显示相关节点。

为了方便实现，修改了Graph里面的tooltip，以及repo-network特性里面的两个class名称

目前存在的问题是，我是使用阈值来控制显示的邻接节点。我找不到Graph使节点高亮的阈值是多少，有许多节点之间是有value，但是并没有在network中高亮出来，我自己找规律找了一些阈值使得表格中仅仅显示高亮的节点。

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/hypertrons/hypertrons-crx?pullRequest=835) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/hypertrons/hypertrons-crx?pullRequest=835) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/hypertrons/hypertrons-crx?pullRequest=835) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/hypertrons/hypertrons-crx?pullRequest=835) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/hypertrons/hypertrons-crx?pullRequest=835) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/hypertrons/hypertrons-crx?pullRequest=835) it.</sub>

wouldmm commented:
具体实现的样例如下图，并且鼠标挪开后能显示原来的文字

![14795cbb-2b3a-4989-9f58-ce9d73eb4d46](https://github.com/user-attachments/assets/6d14b8df-1ebb-41b3-b4ed-a23523cda079)

wxharry commented:
Hi first-time contributor, welcome to the community. I can help with your questions. But it seems like your branch is not building, could you please fix it for me first. Thanks

BTW, Please remember to sign the CLA.

wouldmm commented:
I'm not sure if this error is caused by the mistake that I pulled the yarn.lock file to branch. It seems that some of the code in yarn.lock is updated automatically, and I didn't take the initiative to change it. 

I'm now trying to cancel the yarn.lock changes, and it seems that the pull request automatically synchronizes the change for me. I don't know if I'm doing it correctly or if I need to do a new pull request?

wxharry commented:
> I'm not sure if this error is caused by the mistake that I pulled the yarn.lock file to branch. It seems that some of the code in yarn.lock is updated automatically, and I didn't take the initiative to change it.
> 
> I'm now trying to cancel the yarn.lock changes, and it seems that the pull request automatically synchronizes the change for me. I don't know if I'm doing it correctly or if I need to do a new pull request?

It is fine. yarn.lock should be included, you don't need to worry about it. [read more here](https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored)

wouldmm commented:
I can run the project in my computer, I don't know why there is a build error. It seems to missing file "manifest.json"

wxharry commented on src/pages/ContentScripts/features/network-adj-show/index.tsx:
```diff
@@ -0,0 +1,136 @@
+
+function addTable(adjacentNodes, nodeListSpan) {
+  const table = document.createElement('table');
+  table.style.borderCollapse = 'collapse';
+  table.style.width = '100%';
+
+  // 添加表头
+  const headerRow = document.createElement('tr');
+  const headerKey = document.createElement('th');
+  headerKey.innerText = 'Related Repository';
+  headerKey.style.border = '1px solid white';
+  // headerKey.style.backgroundColor = '#f2f2f2';
+  const headerValue = document.createElement('th');
+  headerValue.innerText = 'Relation value';
+  headerValue.style.border = '1px solid white';
+  // headerValue.style.backgroundColor = '#f2f2f2';
+  headerRow.appendChild(headerKey);
+  headerRow.appendChild(headerValue);
+  table.appendChild(headerRow);
+
+  // 添加数据行
+  adjacentNodes.forEach((item: [string, number]) => {
+    const row = document.createElement('tr');
+    const cellKey = document.createElement('td');
+    cellKey.innerText = item[0];
+    cellKey.style.border = '1px solid white';
+    const cellValue = document.createElement('td');
+    cellValue.innerText = item[1].toString();
+    cellValue.style.border = '1px solid white';
+    row.appendChild(cellKey);
+    row.appendChild(cellValue);
+    table.appendChild(row);
+  });
+
+  // 将表格插入到指定元素中
+  for (const element of nodeListSpan) {
+    element.innerHTML = '';
+    element.appendChild(table);
+  }
+}
```
Can we make this a separate component? It would be easy to maintain.

wouldmm replied:
Actually, I haven't learned much about front-end content, this is just an assignment for the course. I modified the code to present the table in the form of a view and it ran successfully, but there are still warnings when I check the console, and I don't know how to fix that.

thank u for ur help

![image](https://github.com/user-attachments/assets/a5e89bb4-d1f2-4dff-a4b6-b3c6bcd630b7)

wxharry reviewed (changes requested):
I left some comments so we can improve the code. I don't think we need to inject this feature as a extension feature. It can stay at the component level. To be specific, `src/pages/ContentScripts/features/repo-networks/view.tsx`.

We can expose a onToolTipShown function from the graph component. This method will be triggered everytime the tooltip shown and hidden(this can be one boolean), and it also gives what are needed to show the table. Then we can have a table components in the view files and it shows and hides when triggered.

Let me know what you think.

wouldmm commented on src/pages/ContentScripts/features/network-adj-show/index.tsx:
```diff
@@ -0,0 +1,136 @@
+  return adjacentNodes;
+}
+
+function addTable(adjacentNodes, nodeListSpan) {
```
Actually, this error will not affect the program's operation.But I have tried to fix it.

wouldmm commented on src/pages/ContentScripts/features/network-adj-show/index.tsx:
```diff
@@ -0,0 +1,136 @@
+  const con3 = document.getElementsByClassName('color-text-secondary-3');
+  const text3 = con3[0].innerHTML;
+  document.addEventListener('nodeHovered', function (event) {
+    const nodeData = event.detail; // 获取节点数据
```
Actually, this error will not affect the program's operation.But I have tried to fix it.

wouldmm commented on src/components/Graph.tsx:
```diff
@@ -64,7 +64,20 @@ const Graph: React.FC<GraphProps> = ({ data, style = {}, focusedNodeID }) => {
+    tooltip: {
+      trigger: 'item',
+      triggerOn: 'mousemove',
+      formatter: function (params) {
```
Actually, this error will not affect the program's operation.But I have tried to fix it.

wouldmm commented on src/components/Graph.tsx:
```diff
@@ -64,7 +64,20 @@ const Graph: React.FC<GraphProps> = ({ data, style = {}, focusedNodeID }) => {
+        } else {
+          return params.data.source + ' > ' + params.data.target + ' : ' + params.data.value;
+        }
+      },
```
Thanks for ur suggestion.

I followed your suggestion, but I don't know how to use showTip to determine the movement of the mouse because I tried but didn't succeed. Instead, I used mouseover to implement it. I don't know if the code logic for rendering the table is correct because I have never done a similar task before.

wouldmm commented:
Sorry, I accidentally clicked on the resolve conversation, which caused one of your suggestion records to disappear

I followed your suggestion, but I don't know how to use showTip to determine the movement of the mouse because I tried but didn't succeed. On the contrary, I used mouseover to implement it. I don't know if the code logic for rendering the table is correct because I have never done a similar task before.

wxharry commented:
You can leave other issues aside. The first thing to do is to make it react component level. It is not a browser extension level task but a very basic one. Please follow the instructions:
1. add an event handler to Graph.
    a. Add a new prop to GraphProp, it should be a function
    b. Implement the prop in Graph. Call the function whenever a tooltip shows and hides(You can use whatever works for you, either mousehover or tooltipshown). When it is triggered, giving needed params and whether it is shown or hidden. (simliar to what you have already did)
    c. Give a handle function in `src/pages/ContentScripts/features/repo-networks/view.tsx`. You don't need a `index.tsx`, what you need to have a new function and pass it to the Graph element. This function handles the data like you did in the `index.tsx`
 
The expected output for 1) should be like this:

```
// in src/pages/ContentScripts/features/repo-networks/view.tsx
// You have a function to handle data
const onProjectCorrelationNetworkNodeToolTipChange = (params:any)=>{
// parse data here
}

<Graph data={repoNetwork} style={graphStyle} focusedNodeID={currentRepo} 
           onToolTipChange={onProjectCorrelationNetworkNodeToolTipChange} />
```

```
// Graph element has something like this
const Graph: React.FC<GraphProps> = ({ data, style = {}, focusedNodeID, onToolTipChange: onToolTipChange }) => {
```
2. Change graph description visibility
    a. create two new booleans using useState for graphs
    b. Set the variables based on Tooltip change in onProjectCorrelationNetworkNodeToolTipChange
    c. make description show and hide using the variables something like this:
```
const [projectDescriptoinShow, SetProjectDescriptoinShow] = useState(true)
...
const onProjectCorrelationNetworkNodeToolTipChange = (params:any)=>{
SetProjectDescriptoinShow (true/false)
}
...
            { projectDescriptoinShow & 
<div className="color-text-secondary-2" style={{ marginLeft: '35px', marginRight: '35px' }}>
              <p>{t('component_projectCorrelationNetwork_description')}</p>
              <ul style={{ margin: '0px 0 10px 15px' }}>
                <li>{t('component_projectCorrelationNetwork_description_node')}</li>
                <li>{t('component_projectCorrelationNetwork_description_edge')}</li>
              </ul>
            </div>
}
```

3. Have a component to show tables
    a. Add a `Table.tsx` file. To create a component, reference to`src/components/TooltipTrigger/index.tsx` as this is a super simple one. You need to change the name, props, and content.
    b. Add `Table` to the view, pass any parameters needed.

What is expected:
            { projectDescriptoinShow
? <div className="color-text-secondary-2" style={{ marginLeft: '35px', marginRight: '35px' }}>
              <p>{t('component_projectCorrelationNetwork_description')}</p>
              <ul style={{ margin: '0px 0 10px 15px' }}>
                <li>{t('component_projectCorrelationNetwork_description_node')}</li>
                <li>{t('component_projectCorrelationNetwork_description_edge')}</li>
              </ul>
            </div>
: <Table data={tableData} />
}

These are easy code to work on, you just need less than an hour to know all these and another one to make it work. This may be just a course for you but it is not for me. Please show your efforts.

wouldmm commented:
Thank you so much for your patient help. You are really smart, and your suggested method is something I hadn’t thought of. I have done my best to complete this task and hope I won't disappoint you this time.

wxharry commented:
> Thank you so much for your patient help. You are really smart, and your suggested method is something I hadn’t thought of. I have done my best to complete this task and hope I won't disappoint you this time.

Good job this time. Will take a look later. Thanks for you contributions.

wouldmm commented:
Thanks again for your paitent  guidence.

wxharry commented on src/components/Table.tsx:
```diff
@@ -0,0 +1,30 @@
+const Table: React.FC<TableProps> = ({ adjacentNodes }) => {
+  return (
+    <div>
+      <table style={{ borderCollapse: 'collapse', width: '100%' }}>
+        <thead>
+          <tr>
+            <th style={{ border: '1px solid white' }}>Related Repository</th>
+            <th style={{ border: '1px solid white' }}>Relation value</th>
+          </tr>
+        </thead>
+        <tbody>
+          {adjacentNodes.map((node, index) => (
+            <tr key={index}>
+              <td style={{ border: '1px solid white', textAlign: 'center', verticalAlign: 'middle' }}>{node[0]}</td>
+              <td style={{ border: '1px solid white', textAlign: 'center', verticalAlign: 'middle' }}>{node[1]}</td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
```
We need to change the table style based on Lam's desgin.
<img width="265" alt="image" src="https://github.com/user-attachments/assets/069a8a89-510b-43e9-9504-ef6c78d644fe">
1. Remove the border color
2. text in the first column should be left align
3. We need to add avatars (This is important because users need avatars to matching highlighted nodes in graph and names in table)
4. We need to handle the situation when there are too many to show. Add a <div> as a table container and set its max height and overflow.
<img width="1244" alt="image" src="https://github.com/user-attachments/assets/17b9a4d1-f798-408c-bd02-fa10cf6e5939">
5. You can have a variable to hold the same style instead of setting it repeatedly.
6. We also need margins for the tables

wxharry commented on src/components/Graph.tsx:
```diff
@@ -115,6 +120,23 @@ const Graph: React.FC<GraphProps> = ({ data, style = {}, focusedNodeID }) => {
+      });
+      instance.on('mouseout', function (param: any) {
+        if (param.dataType === 'node') {
+          const nodeDataIndex = param.data.id;
+          if (onToolTipChange) {
+            onToolTipChange(nodeDataIndex, false);
+          }
```
This part can be simplified into one single line:  `onToolTipChange && onToolTipChange(param.data.id, false);` Same for the mouseover.

wxharry commented on src/components/Graph.tsx:
```diff
@@ -17,6 +17,11 @@ interface GraphProps {
   readonly focusedNodeID?: string;
+
+  /**
+   * callback function when mouse hover node
```
Let's improve this comment, it is called when mouse hover our and out on a node data.

wxharry reviewed (changes requested):
Sorry for the delay. Nicely done! I added some comments and most of them are minor issues. We still need to work on the table a bit more, but I am sure we are almost done here.

wouldmm commented:
I modified the code according to your suggestion. In addition, I considered that if the table is too long and I want to view more complete node data, the table will disappear when the mouse moves out of the graph. My current approach is that when the mouse moves into the right component, the table will reappear. And I added a hyperlink to the data of each node. I don't know if this solution is reasonable.

wxharry commented:
> I modified the code according to your suggestion. In addition, I considered that if the table is too long and I want to view more complete node data, the table will disappear when the mouse moves out of the graph. My current approach is that when the mouse moves into the right component, the table will reappear. And I added a hyperlink to the data of each node. I don't know if this solution is reasonable.

Not sure this is a good design. cc @HalloMelon. Do you have any idea how to properly handle this problem?

wxharry commented:
@wouldmm Let's add an icon and a tooltip for description and show table on mouse hovering. The table should use current repo as by default(for the first time we open the perceptor tab, show the table for hovering on the current repo node)

<img width="841" alt="image" src="https://github.com/user-attachments/assets/730aede9-c999-4e1a-9ea5-f466861019de">

You can find the code in options page.

I noticed you use links as repo name, it is useful. Well done! Should be able to merge after this change.

wouldmm commented:
I tried to add a tooltip to display text, which was basically implemented, but after adding it, the graph moved to the left. I tried many ways but still couldn't fix this problem. And when I press F12 to view the web developer, the graph suddenly jumps to the center. Sorry, I don't know what to do.

wxharry commented:
> I tried to add a tooltip to display text, which was basically implemented, but after adding it, the graph moved to the left. I tried many ways but still couldn't fix this problem. And when I press F12 to view the web developer, the graph suddenly jumps to the center. Sorry, I don't know what to do.

Looks like it is because of the Popover. Replace it with NativePopover fix this problem.

wouldmm commented:
I‘ve changed the Popover to NativePopover.
